plutus-metatheory/src/VerifiedCompilation/UCSE.lagda.md

```diff
@@ -0,0 +1,63 @@
+---
+title: VerifiedCompilation.UCSE
+layout: page
+---
+
+# Common Subexpression Elimination Translation Phase
+```
+module VerifiedCompilation.UCSE where
+
+```
+## Imports
+
+```
+open import VerifiedCompilation.Equality using (DecEq; _≟_; decPointwise)
+open import VerifiedCompilation.UntypedViews using (Pred; isCase?; isApp?; isLambda?; isForce?; isBuiltin?; isConstr?; isDelay?; isTerm?; allTerms?; iscase; isapp; islambda; isforce; isbuiltin; isconstr; isterm; allterms; isdelay)
+open import VerifiedCompilation.UntypedTranslation using (Translation; translation?; Relation)
+open import Relation.Nullary.Product using (_×-dec_)
+open import Data.Product using (_,_)
+import Relation.Binary as Binary using (Decidable)
+open import Relation.Nullary using (Dec; yes; no; ¬_)
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+import Relation.Binary.PropositionalEquality as Eq
+open Eq using (_≡_; refl)
+open import Data.Empty using (⊥)
+open import Agda.Builtin.Maybe using (Maybe; just; nothing)
+open import Untyped.RenamingSubstitution using (_[_])
+```
+## Translation Relation
+
+This module is required to certify that an application of CSE doesn't break the
+semantics; we are explicitly not evaluating whether the particular choice of
+sub-expression was a "good" choice.
+
+As such, this Translation Relation primarily checks that substituting the expression
+back in would yield the original expression.
+
+```
+data UCSE : Relation where
+  cse : {X : Set} {x : Maybe X ⊢} {x' e : X ⊢}
+    → Translation UCSE (x [ e ]) x'
+    → UCSE ((ƛ x) · e) x'
+
+UntypedCSE : {X : Set} {{_ : DecEq X}} → (ast : X ⊢) → (ast' : X ⊢) → Set₁
+UntypedCSE = Translation UCSE
+
+```
+
+## Decision Procedure
+
+```
+
+isUntypedCSE? : {X : Set} {{_ : DecEq X}} → Binary.Decidable (Translation UCSE {X})
+
+{-# TERMINATING #-}
+isUCSE? : {X : Set} {{_ : DecEq X}} → Binary.Decidable (UCSE {X})
+isUCSE? ast ast' with (isApp? (isLambda? isTerm?) isTerm?) ast
+... | no ¬match = no λ { (cse x) → ¬match (isapp (islambda (isterm _)) (isterm _)) }
+... | yes (isapp (islambda (isterm x)) (isterm e)) with isUntypedCSE? (x [ e ]) ast'
```

Comment:
I think the whole approach to certification is very suboptimal. I.e. this pass and the others (not sure if all of them). So what you do here is you check that the LHS has the right shape and then you proceed with substitution and recurse. Potentially just to figure out that you didn't need to do all this work, because the terms match exactly and there was no point in all these substitutions and non-deterministic digressing into CSE, float-delay and everything else at every lambda application node.

I guess this is a fine start, but I wouldn't be surprised if it eventually turns out that all the passes need to be rewritten, because they can't handle real-world terms fast enough.

Reply:
If they are identical then that should be handled by `translation?` before it gets into this decision procedure. 

It isn't because this one runs first in `translation?` because if it didn't then we might get half way down the AST before realising we need to look at some of the higher structure for the specific relation, and I'm not sure how neat the backtracking is in `translation?`. That is what I'd look at first...

Reply:
> If they are identical then that should be handled by `translation?` before it gets into this decision procedure.

Nope, `translation?` invokes `isR?` before doing anything else.

Reply:
The word "should" was important in that sentence :)